# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](tttboard.jpg)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [3]:
# Read tic-tac-toe.csv into a dataframe
import pandas as pd

df = pd.read_csv('tic-tac-toe.csv')  
print("Data Loaded Successfully!")
print(df.head())  # Preview the first few rows

# Inspect the dataset
print("\nDataset Info:")
print(df.info())  # Check data types and non-null counts

print("\nValue counts of class column:")
print(df['class'].value_counts())  # Check class balance

# Convert categorical values to numeric
df = df.replace({'x': 1, 'o': 0, 'b': -1, 'positive': 1, 'negative': 0})

# Separate inputs and outputs
X = df.drop('class', axis=1)
y = df['class']

# Normalize the input data (already numerical but i will scale between 0 and 1)
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Print shape to verify
print("\nShape of features:", X_scaled.shape)
print("Shape of labels:", y.shape)


Data Loaded Successfully!
  TL TM TR ML MM MR BL BM BR  class
0  x  x  x  x  o  o  x  o  o   True
1  x  x  x  x  o  o  o  x  o   True
2  x  x  x  x  o  o  o  o  x   True
3  x  x  x  x  o  o  o  b  b   True
4  x  x  x  x  o  o  b  o  b   True

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   TL      958 non-null    object
 1   TM      958 non-null    object
 2   TR      958 non-null    object
 3   ML      958 non-null    object
 4   MM      958 non-null    object
 5   MR      958 non-null    object
 6   BL      958 non-null    object
 7   BM      958 non-null    object
 8   BR      958 non-null    object
 9   class   958 non-null    bool  
dtypes: bool(1), object(9)
memory usage: 68.4+ KB
None

Value counts of class column:
class
True     626
False    332
Name: count, dtype: int64


/var/folders/2n/gs_200v517s5clp2y1jt0mgc0000gn/T/ipykernel_18853/2895840015.py:16: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.replace({'x': 1, 'o': 0, 'b': -1, 'positive': 1, 'negative': 0})



Shape of features: (958, 9)
Shape of labels: (958,)


## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [4]:
# For data handling
import numpy as np
from sklearn.model_selection import train_test_split

# For building the model
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [5]:
# Split into 80% train and 20% test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Confirm dimensions
print("Training features shape:", X_train.shape)
print("Test features shape:", X_test.shape)

Training features shape: (766, 9)
Test features shape: (192, 9)


In [6]:
# Define the model
model = Sequential()
model.add(Dense(32, activation='relu', input_shape=(9,)))
model.add(Dense(64, activation='relu'))
model.add(Dense(2, activation='softmax'))  # 2 output classes: win or not win

/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
# Compile the model
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [8]:
history = model.fit(X_train, y_train, epochs=50, validation_split=0.1)

Epoch 1/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6471 - loss: 0.6486 - val_accuracy: 0.7143 - val_loss: 0.6051
Epoch 2/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6690 - loss: 0.6042 - val_accuracy: 0.7013 - val_loss: 0.5938
Epoch 3/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6683 - loss: 0.6061 - val_accuracy: 0.6883 - val_loss: 0.5833
Epoch 4/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6937 - loss: 0.5640 - val_accuracy: 0.7143 - val_loss: 0.5709
Epoch 5/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7210 - loss: 0.5631 - val_accuracy: 0.7143 - val_loss: 0.5609
Epoch 6/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7078 - loss: 0.5680 - val_accuracy: 0.7273 - val_loss: 0.5478
Epoch 7/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7431 - loss: 0.5279 - val_accuracy: 0.7143 - val_loss: 0.5411
Epoch 8/50
22/22 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7480 - loss: 0.5161 - val_accuracy: 0.7532 - val_loss:

In [9]:
# Evaluate on test set
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"Test Accuracy: {test_accuracy:.4f}")

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8464 - loss: 0.4031
Test Accuracy: 0.8594


In [11]:
# Save model to file
model.save("tic-tac-toe.model.keras")
print("Model saved successfully.")

Model saved successfully.


## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [15]:
# Load the saved model
from tensorflow.keras.models import load_model

model = load_model("tic-tac-toe.model.keras")
print("Model loaded!")

# Make predictions on random test samples
import numpy as np

# Picking a few random samples
sample_indices = np.random.choice(len(X_test), size=5, replace=False)

# Convert to NumPy arrays first
X_test_np = X_test.to_numpy()
y_test_np = y_test.to_numpy()

sample_inputs = X_test_np[sample_indices]
sample_labels = y_test_np[sample_indices]


# Run model predictions
predictions = model.predict(sample_inputs)

# Get predicted classes (index of highest probability)
predicted_classes = np.argmax(predictions, axis=1)

# Display predictions vs actuals
for i in range(len(sample_indices)):
    print(f"Sample {i+1}")
    print("Predicted:", predicted_classes[i])
    print("Actual:   ", sample_labels[i])
    print("---")


Model loaded!
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step
Sample 1
Predicted: 0
Actual:    False
---
Sample 2
Predicted: 0
Actual:    False
---
Sample 3
Predicted: 1
Actual:    True
---
Sample 4
Predicted: 1
Actual:    True
---
Sample 5
Predicted: 1
Actual:    True
---


## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

In [ ]:
model = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(9,)),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dense(2, activation='softmax')
])

/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [21]:
# Adjust the learning rate
optimizer = tf.keras.optimizers.Adam(learning_rate=0.0005)

In [22]:
# Recompile the model and retrain
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(X_train, y_train, epochs=20, validation_split=0.2)

Epoch 1/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5133 - loss: 0.7042 - val_accuracy: 0.5714 - val_loss: 0.6672
Epoch 2/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5612 - loss: 0.6993 - val_accuracy: 0.5844 - val_loss: 0.6487
Epoch 3/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5981 - loss: 0.6680 - val_accuracy: 0.6104 - val_loss: 0.6368
Epoch 4/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5846 - loss: 0.6699 - val_accuracy: 0.6234 - val_loss: 0.6304
Epoch 5/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6369 - loss: 0.6382 - val_accuracy: 0.6364 - val_loss: 0.6225
Epoch 6/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6500 - loss: 0.6333 - val_accuracy: 0.6234 - val_loss: 0.6187
Epoch 7/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6310 - loss: 0.6478 - val_accuracy: 0.6299 - val_loss: 0.6147
Epoch 8/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6505 - loss: 0.6294 - val_accuracy: 0.6494 - val_loss:

In [23]:
# Rebuild and train the improved model (with same 20 epochs)
from tensorflow.keras.optimizers import Adam

# Build the model
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(9,)))   # 1st hidden layer
model.add(Dense(32, activation='relu'))                     # 2nd hidden layer
model.add(Dense(2, activation='softmax'))                   # output layer

# Compile with custom learning rate
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Fit model (20 epochs, as before)
history = model.fit(X_train, y_train, epochs=20, validation_data=(X_test, y_test))


Epoch 1/20


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5095 - loss: 0.7059 - val_accuracy: 0.6562 - val_loss: 0.6464
Epoch 2/20
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6647 - loss: 0.6052 - val_accuracy: 0.6823 - val_loss: 0.6159
Epoch 3/20
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6985 - loss: 0.5901 - val_accuracy: 0.7031 - val_loss: 0.5948
Epoch 4/20
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7659 - loss: 0.5347 - val_accuracy: 0.7135 - val_loss: 0.5749
Epoch 5/20
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7706 - loss: 0.5296 - val_accuracy: 0.7396 - val_loss: 0.5601
Epoch 6/20
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7674 - loss: 0.5128 - val_accuracy: 0.7292 - val_loss: 0.5454
Epoch 7/20
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7611 - loss: 0.5196 - val_accuracy: 0.7552 - val_loss: 0.5300
Epoch 8/20
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7758 - loss: 0.4804 - val_accuracy: 0.7812 - val_loss: 0.5106
Epo

In [24]:
# Evaluate the model on the test dataset
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"Final Test Accuracy: {test_accuracy:.4f}")
print(f"Final Test Loss: {test_loss:.4f}")

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8282 - loss: 0.4113 
Final Test Accuracy: 0.8229
Final Test Loss: 0.4099


**Which approach(es) did you find helpful to improve your model performance?**

In my attempts to improve the model, I tried adding more hidden layers, reducing the learning rate,
and adjusting the number of training epochs. However, these changes did not result in a better model.

The first model (Run 1) performed the best, with a validation accuracy of 0.92 and low validation loss around 0.25. 
In contrast, the second run performed worse, with a maximum validation accuracy of 0.82 and a higher validation loss.

This shows that sometimes, a simpler architecture with enough training time can outperform a more complex one. 
Overfitting was not a major issue in the first run, so reducing training epochs or tweaking regularization didn’t help.